### PR TITLE
core(main-resource): fix protocol error when page is reloaded

### PR DESCRIPTION
**Pull request by connorjclark**
Fixes https://github.com/GoogleChrome/lighthouse/issues/10876

tldr `location.reload` during a navigation results in `MainDocumentContent` artifact trying to fetch the contents of a resource that Chrome backend has already evicted. We only use this artifact in one place `charset`, FYI.

`mainDocumentUrl` is defined to be the last document during a navigation, so we should respect that in `main-resource` by returning the last record that matches this url. Note, the smoke test provided in this PR _did not_ error (w/o these changes) if the page JS-navigates to a different url (ex: `?redirect`), because in that case only one record applies and the current code correctly selects it.

```
Co-authored-by: Wojciech Reszelewski <wreszelewski@gmail.com>
```

### CODE REVIEW

File: core/computed/main-resource.js [hunk 1]
```diff
@@ -22,7 +22,16 @@ class MainResource {
     const {mainDocumentUrl} = data.URL;
     if (!mainDocumentUrl) throw new Error('mainDocumentUrl must exist to get the main resource');
     const requests = await NetworkRecords.request(data.devtoolsLog, context);
-    const mainResource = NetworkAnalyzer.findResourceForUrl(requests, mainDocumentUrl);
+
+    const mainResourceRequests = requests.filter(request =>
```
Comment (connorjclark) [inline core/computed/main-resource.js:26]: 1- Should we add a new function to `NetworkAnalyzer` for `findLastResourceForUrl`?

2- I got rid of the `resourceUrl.startsWith(request.url)` optimization check because my test case tripped up how it may be broken. I think it might need to become `(resourceUrl.startsWith(request.url) || request.url.startsWith(resourceUrl))` to be correct?

Comment (adamraine) [reply core/computed/main-resource.js:26]: > Should we add a new function to NetworkAnalyzer for findLastResourceForUrl?

I think this is a good idea.

> I got rid of the resourceUrl.startsWith(request.url) optimization check because my test case tripped on it, suggesting that it may be broken.

What did it trip on?

File: core/computed/main-resource.js [hunk 2]
```diff
@@ -22,7 +22,16 @@ class MainResource {
...
+      UrlUtils.equalWithExcludedFragments(request.url, mainDocumentUrl)
+    );
+    // We could have more than one record matching the main doucment url,
+    // if the page did `location.reload()`. Since `mainDocumentUrl` refers to the _last_
+    // document request, we should return the last candidate here. Besides, the browser
+    // would have evicted the first request by the time `MainDocumentRequest` (a consumer
+    // of this computed artifact) attempts to fetch the contents, resulting in a protocol error.
+    const mainResource = mainResourceRequests[mainResourceRequests.length - 1];
```
Comment (adamraine) [inline core/computed/main-resource.js:34]: There are a couple other places where we get the main document request this way:
https://github.com/GoogleChrome/lighthouse/blob/1b843de669e18c144422561010985d0dde05d2c6/core/lib/navigation-error.js#L119
https://github.com/GoogleChrome/lighthouse/blob/1b843de669e18c144422561010985d0dde05d2c6/core/computed/page-dependency-graph.js#L409-L414
Should we be getting the last request in these places as well?

**Comment by brendankenny** [comment]
The `update:sample-json` errors are correct: you can't assume the last request for the main document URL is the actual main document request (thanks dbw for being such a gross site that you'd XHR your own html :)

Is this case generally recoverable? There's an unexpected navigation in the middle of the navigation.

**Comment by connorjclark** [comment]
Hehe, this is why sample json's critical request chain broke:

image

For dbw_tester.html, that xhr request is now the main resource! Just need to check the `type` is `Document` :)

**Comment by connorjclark** [comment]
> I guess you can think of it as a redirect, but there's going to be a lot of strange stuff going on vs a typical redirect, resources already in the cache, etc

This seems better than the alternatives:

1) catch this error and make the charset audit n/a
2) re-define what MainDocumentContent/mainDocumentUrl means, if we don't really mean "the last document request"

We could hedge things a bit by emitting a warning when we detect a reload, treating it like the redirect warning.

File: core/computed/main-resource.js (hunk 1 above)
Comment (connorjclark) [reply core/computed/main-resource.js:26]: Two urls can be equal ignoring the fragment, without one being a substring of the other.

Comment (brendankenny) [reply core/computed/main-resource.js:26]: `request.url` won't ever have fragments, so it should always be the start of a URL that it's equal to when excluding fragments.

**Comment by brendankenny** [comment]
iframes of itself will also be caught by this but have type Document.

The real alternative is issuing a navigationError in this case. It doesn't seem like any metrics should be trusted for this kind of load.

**Comment by adamraine** [comment]
> iframes of itself will also be caught by this but have type Document.

Would the [initiator type](https://chromedevtools.github.io/devtools-protocol/tot/Network/#type-Initiator) help us in this case? Looks like the main document has type `other` when iframes have type `parser`

**Comment by wreszelewski** [comment]
Can I help somehow to push this PR further?

File: core/computed/main-resource.js [hunk 3]
```diff
@@ -21,8 +21,14 @@ class MainResource {
...
-    const requests = await NetworkRecords.request(data.devtoolsLog, context);
-    const mainResource = NetworkAnalyzer.findResourceForUrl(requests, mainDocumentUrl);
+    const records = await NetworkRecords.request(data.devtoolsLog, context);
+
+    // We could have more than one record matching the main doucment url,
+    // if the page did `location.reload()`. Since `mainDocumentUrl` refers to the _last_
+    // document request, we should return the last candidate here. Besides, the browser
+    // would have evicted the first request by the time `MainDocumentRequest` (a consumer
+    // of this computed artifact) attempts to fetch the contents, resulting in a protocol error.
+    const mainResource = NetworkAnalyzer.findLastDocumentForUrl(records, mainDocumentUrl);
```
Comment (adamraine) [inline core/computed/main-resource.js:31]: I think it makes sense to do this in `page-dependency-graph.js` as well. WDYT?

**Review by adamraine — APPROVED** [review]
